huaweicloud/services/waf/resource_huaweicloud_waf_dedicated_domain.go

```diff
@@ -238,18 +259,71 @@ func resourceWafDedicatedDomainV1Create(d *schema.ResourceData, meta interface{}
 			logp.Printf("[ERROR] error change the protection status of WAF dedicate domain %s: %s", d.Id(), err)
 		}
 	}
-	if v, ok := d.GetOk("tls"); ok {
-		updateOpts := domains.UpdatePremiumHostOpts{
-			Tls: v.(string),
+
+	if err := updateWafDedicatedDomain(wafDedicatedClient, meta, d); err != nil {
+		return fmtp.Errorf("error updating WAF dedicated domain: %s", err)
+	}
+
+	return resourceWafDedicatedDomainV1Read(d, meta)
+}
+
+func updateWafDedicatedDomain(client *golangsdk.ServiceClient, meta interface{}, d *schema.ResourceData) error {
+	updateOpts := domains.UpdatePremiumHostOpts{}
+	if d.HasChange("tls") {
+		updateOpts.Tls = d.Get("tls").(string)
+	}
+
+	if d.HasChange("cipher") {
+		updateOpts.Cipher = d.Get("cipher").(string)
+	}
+
+	if d.HasChange("proxy") {
+		proxy := d.Get("proxy").(bool)
+		updateOpts.Proxy = &proxy
+	}
+
+	if d.HasChange("certificate_id") {
+		if v, ok := d.GetOk("certificate_id"); ok {
+			certName, err := getCertificateNameById(d, meta)
+			if err != nil {
+				return err
+			}
+			updateOpts.CertificateName = certName
+			updateOpts.CertificateId = v.(string)
 		}
-		logp.Printf("[DEBUG] Waf dedicated domain set TLS: %#v", updateOpts)
-		_, err = domains.Update(wafDedicatedClient, d.Id(), updateOpts)
+	}
+
+	if d.HasChanges("pci_3ds", "pci_dss") {
+		flag, err := getHostFlag(d)
 		if err != nil {
-			return fmtp.Errorf("error updating WAF dedicated domain TLS: %s", err)
+			return err
 		}
+		updateOpts.Flag = flag
 	}
+	logp.Printf("[DEBUG] Waf dedicated domain update: %#v", updateOpts)
+	_, err := domains.Update(client, d.Id(), updateOpts)
```

Comment:
what will happen if updateOpts is empty?

Reply:
The api response will be ok and nothing will be changed if updateOpts is empty.
I will add preCheck to make sure updateOpts is not empty.
```
for create

if d.HasChanges("tls", "cipher", "pci_3ds", "pci_dss") {
}
```

```
for update

if d.HasChanges("tls", "cipher", "proxy", "certificate_id", "pci_3ds", "pci_dss") {
}
```